language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -127,6 +139,113 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
         existing_borrows.is_empty()
     }
 
+    fn verify_instantiation(
+        env: (&[StructHandle], &[Kind], usize),
+        tys: &[SignatureToken],
+        type_formals: &[Kind],
+    ) -> Result<Vec<Kind>, VMStaticViolation> {
+        let (_, _, offset) = env;
+
+        // Check if the number of formals and the number of actuals match.
+        if type_formals.len() != tys.len() {
+            return Err(VMStaticViolation::NumberOfTypeActualsMismatch(
+                offset,
+                type_formals.len(),
+                tys.len(),
+            ));
+        }
+
+        // Recursively verify the type actuals and gather their kinds.
+        let kinds = tys
+            .iter()
+            .map(|ty| Self::verify_type(env, ty))
+            .collect::<Result<Vec<_>, VMStaticViolation>>()?;
+
+        // Check if each kind is a sub-kind of that specified in the constraint.
+        if kinds
+            .iter()
+            .zip(type_formals.iter())
```

Comment:
I don't think you need the `.iter()` here, just do `type_formals`
`zip` takes anything that implements `IntoIterator`

Reply:
Nope. Doesn't compile.